R/checkout.R

```diff
@@ -28,64 +55,142 @@ checkout <- function(repos = getOption("repos"),
   renv_project_lock(project = project)
 
   # set new repositories
+  repos <- repos %??% renv_checkout_repos(date)
   options(repos = repos)
 
+  # TODO: Activate Bioconductor if it appears to be used by this project
+
   # select packages to install
-  packages <- packages %||% renv_checkout_packages(project = project)
-  names(packages) <- packages
+  packages <- packages %??% renv_checkout_packages(project = project)
+
+  # get the associated remotes for these packages
+  remotes <- renv_checkout_remotes(packages, project)
+
+  # parse these into package records
+  records <- map(remotes, renv_remotes_resolve)
+
+  # create a lockfile matching this request
+  lockfile <- renv_lockfile_init(project)
+  lockfile$Packages <- records
+
+  # restore from that lockfile
+  restore(lockfile = lockfile, clean = clean)
+
+}
+
+renv_checkout_packages <- function(project) {
+  deps <- dependencies(project, progress = FALSE)
+  unique(deps$Package)
+}
+
+renv_checkout_remotes <- function(packages, project) {
 
-  # install all of these packages, effectively attempting to update
-  # to the latest version of each package
-  latest <- lapply(packages, function(package) {
+  # keep only packages which appear to be available in the repositories
+  dbs <- available_packages(type = "source")
+  if (is.null(dbs))
+    stop("no package repositories are available")
 
-    # ignore base packages (these are never distributed on CRAN)
-    priority <- renv_package_priority(package)
-    if (identical(priority, "base"))
-      return(NULL)
+  db <- renv_available_packages_flatten(dbs)
+  packages <- intersect(packages, db$Package)
```

Comment:
I think it would be useful to message `setdiff(db$Package, packages)` so it's clear what's not updated.

Reply:
FWIW you do get a summary of the changes; e.g. if `pak` integration is enabled:

```
> renv::checkout(date = "2022-01-02")
* Snapshot date '2022-01-02' not available; using '2021-12-29' instead
* Querying repositories for available source packages ... Done!
✔ Updated metadata database: 2.73 MB in 5 files.                                              
ℹ R 4.3 aarch64-apple-darwin20 packages are missing from Bioconductor                          
✔ Updating metadata database ... done                                                         
                                                                                                
→ Will install 12 packages.
→ Will update 121 packages.
→ Will download 5 CRAN packages (4.69 MB), cached: 128 (0 B).
+ assertthat                0.2.1    ⬇ (53.25 kB)
+ BiocManager    1.30.20  → 1.30.16 👷🏽
+ bit            4.0.5    → 4.0.4   👷🏿‍♀️🔧
+ blob           1.2.4    → 1.2.2   👷🏿‍♂️
+ boot           1.3-28.1 → 1.3-28  👷🏼‍♀️
+ brew           1.0-8    → 1.0-6   👷🏿‍♀️
+ broom          1.0.4    → 0.7.10  👷🏻
< ... >
```

Do you think that's sufficient? Or should we explicitly call out packages which are installed from GitHub (or other remote sources) but don't appear to be available from the package repositories?